src/pybind/mgr/cephadm/services/ingress.py

```diff
@@ -225,7 +225,7 @@ def keepalived_generate_config(
 
         # set state. first host in placement is master all others backups
         state = 'BACKUP'
-        if hosts[0] == host:
+        if hosts and hosts[0] == host:
```

Comment:
I am not sure if that is the best solution, my reasoning is as follows:
If no demons are running why should we start up something that will grab the ip preventing hosts with working haproxy taking it (another BACKUP with lower priority).

This seem to happen when haproxy is not running at any hosts.

Could we get away with:
```
diff --git a/src/pybind/mgr/cephadm/services/ingress.py b/src/pybind/mgr/cephadm/services/ingress.py
index 472a3c29d2e..34a69d3dede 100644
--- a/src/pybind/mgr/cephadm/services/ingress.py
+++ b/src/pybind/mgr/cephadm/services/ingress.py
@@ -187,6 +187,9 @@ class IngressService(CephService):
 
         host = daemon_spec.host
         hosts = sorted(list(set([str(d.hostname) for d in daemons])))
+        if len(hosts) == 0:
+            raise OrchestratorError(
+                f"Unable to find any haproxy daemons for service {spec.service_name()}, not starting keepalived to grab ip.")
```

Maybe a bit off topic, but, why are we using the host of the haproxy service to decide if it is MASTER or BACKUP, would it not make more sense to either:
1. use the spec to decide directly (first host of placement hosts)
2. multi-master

Reply:
That would work for me :)

As the keepalived config is now, the master (with the lowest ip) will preempt the ip back when/if it comes back. I guess most people would not really care as long as something has the ip running with a haproxy instance and some people would want to balance bandwidth between nodes (are they more likely to use external loadbalancers?).

I wonder if piggybacking on ceph cluster quorum to make decisions on who grabs the ip would be better than using VRRP. In a bad state we could end up with a VRRP master in the non-quorum part of the cluster, (which still may be something you want :shrug:). And now I am writing an essay in a comment :smile: Just food for future thoughts ..